### PR TITLE
[Futures] Simulated future support

### PR DESCRIPTION
**Pull request by Herklos**

### CODE REVIEW

**Comment by valouvaliavlo** [comment]
👍

**Comment by valouvaliavlo** [comment]
If a second position(same side) is open, it'll result in one mean of other ?

**Comment by Herklos** [comment]
> If a second position(same side) is open, it'll result in one mean of other ?

It's not possible to open a new position on the same side of the first one (even if in hedge position mode). You have to close the first one before.

**Comment by valouvaliavlo** [comment]
> > If a second position(same side) is open, it'll result in one mean of other ?
> 
> It's not possible to open a new position on the same side of the first one (even if in hedge position mode). You have to close the first one before. 

You can place another orders, when it's taken it'll DCA your position

**Comment by Herklos** [comment]
> > > If a second position(same side) is open, it'll result in one mean of other ?
> > 
> > 
> > It's not possible to open a new position on the same side of the first one (even if in hedge position mode). You have to close the first one before.
> 
> You can place another orders, when it's taken it'll DCA your position

A new order will increase or decrease the position size depending of the side.

File: octobot_trading/personal_data/portfolios/assets/future_asset.pxd [hunk 1]
```diff
@@ -18,15 +18,17 @@ cimport octobot_trading.personal_data.portfolios.asset as asset_class
...
+    cdef public object order_margin
 
-    cpdef bint update(self, object available=*, object total=*, object initial_margin=*, object margin_balance=*,
-                      object maintenance_margin=*, object position_initial_margin=*, object unrealised_pnl=*)
-    cpdef bint set(self, object available=*, object total=*, object initial_margin=*, object margin_balance=*,
-                   object maintenance_margin=*, object position_initial_margin=*, object unrealised_pnl=*)
-    cdef void _update_equity(self)
+    cpdef bint update(self, object total=*, object available=*, object position_margin=*,
+                      object unrealized_pnl=*, object initial_margin=*)
+    cpdef bint set(self, object total=*, object available=*, object margin_balance=*, object initial_margin=*,
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/portfolios/assets/future_asset.pxd:29]: by returning bint here and in `update`, we will block potential raised exceptions, it's not a problem here ?

File: octobot_trading/personal_data/portfolios/portfolio.py [hunk 2]
```diff
@@ -168,83 +160,51 @@ def get_portfolio_from_amount_dict(self, amount_dict):
...
             raise RuntimeError("Portfolio has to be initialized using decimal.Decimal")
-        return {currency: self._create_currency_portfolio(available=total, total=total)
+        return {currency: self.create_currency_asset(currency=currency, available=total, total=total).to_dict()
                 for currency, total in amount_dict.items()}
 
     def __str__(self):
         return f"{personal_data.portfolio_to_float(self.portfolio)}"
 
-    def _update_portfolio_data(self, currency, value, total=True, available=False):
+    def _update_portfolio_data(self, currency, total_value=constants.ZERO, available_value=constants.ZERO,
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/portfolios/portfolio.py:169]: 👍 much clearer

**Review by GuillaumeDSM — COMMENTED** [review]
👍 
(I'm not finished yet, I'm at https://github.com/Drakkar-Software/OctoBot-Trading/pull/656/files#diff-c2c76ae665246251d9788bc627c648b64dbc552e3229031e9373d84995a0b877L36)

**Comment by valouvaliavlo** [comment]
> > > > If a second position(same side) is open, it'll result in one mean of other ?
> > > 
> > > 
> > > It's not possible to open a new position on the same side of the first one (even if in hedge position mode). You have to close the first one before.
> > 
> > 
> > You can place another orders, when it's taken it'll DCA your position
> 
> A new order will increase or decrease the position size depending of the side.

Two position with same side and same symbol should be merged in one (quantity should be additionnated and price is an average according to the quantity)

**Comment by Herklos** [comment]
> > > > > If a second position(same side) is open, it'll result in one mean of other ?
> > > > 
> > > > 
> > > > It's not possible to open a new position on the same side of the first one (even if in hedge position mode). You have to close the first one before.
> > > 
> > > 
> > > You can place another orders, when it's taken it'll DCA your position
> > 
> > 
> > A new order will increase or decrease the position size depending of the side.
> 
> Two position with same side and same symbol should be merged in one (quantity should be additionnated and price is an average according to the quantity)

Yes that's what I'm saying and that's what is implemented in this PR.

File: octobot_trading/personal_data/portfolios/assets/future_asset.pxd (hunk 1 above)
Comment (Herklos) [reply octobot_trading/personal_data/portfolios/assets/future_asset.pxd:29]: Yes it's an issue for now. I tried to return object here and in multiple intern method to solve the issue but it didn't work.

File: octobot_trading/personal_data/portfolios/asset.pxd [hunk 3]
```diff
@@ -26,3 +26,4 @@ cdef class Asset:
     cpdef dict to_dict(self)
 
     cpdef object _ensure_update_validity(self, object origin_quantity, object update_quantity)
+    cpdef object _ensure_not_negative(self, object new_value, object replacement_value=*)
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/personal_data/portfolios/asset.pxd:29]: It's an equivalent of "protected" methods. I don't find an equivalent of it

File: octobot_trading/personal_data/orders/order_util.py [hunk 4]
```diff
@@ -101,20 +101,28 @@ async def get_pre_order_data(exchange_manager, symbol: str, timeout: int = None)
...
+        .get_currency_portfolio(market).available
+
+    if exchange_manager.is_future:
+        pair_future_contract = exchange_manager.exchange.get_pair_future_contract(symbol)
+        if pair_future_contract.is_inverse_contract():
+            currency_available *= pair_future_contract.current_leverage
+            market_quantity = market_available = currency_available
+        else:
+            market_available *= pair_future_contract.current_leverage / mark_price
+            market_quantity = currency_available = market_available
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/personal_data/orders/order_util.py:120]: Nope, that's not the same assigment :wink:

File: octobot_trading/personal_data/portfolios/assets/future_asset.pxd (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/portfolios/assets/future_asset.pxd:29]: What was the problem with returning object ? 😮

Comment (Herklos) [reply octobot_trading/personal_data/portfolios/assets/future_asset.pxd:29]: This was not a problem, it was not solving the issue

File: tests/exchanges/traders/test_trader.py [hunk 5]
```diff
@@ -678,14 +678,17 @@ async def test_sell_all_currencies(self):
...
+            name="ADA",
+            available=decimal.Decimal("1500"),
+            total=decimal.Decimal("1500")
+        )
+
+        portfolio_manager.portfolio.portfolio["USDT"] = portfolio_assets.SpotAsset(
+            name="USDT",
+            available=decimal.Decimal("1000"),
+            total=decimal.Decimal("1000")
+        )
```
Comment (GuillaumeDSM) [inline tests/exchanges/traders/test_trader.py:691]: 👍

File: tests/modes/test_abstract_mode_consumer.py [hunk 6]
```diff
@@ -163,11 +162,9 @@ async def test_get_holdings_ratio():
...
     # let channel register proceed
     await wait_asyncio_next_cycle()
-    assert round(ratio, 8) == 1
-    ratio = await consumer.get_holdings_ratio("USDT")
-    assert round(ratio, 8) == 0
-    ratio = await consumer.get_holdings_ratio("XYZ")
-    assert round(ratio, 8) == 0
+    assert await consumer.get_holdings_ratio("BTC") == constants.ONE
+    assert await consumer.get_holdings_ratio("USDT") == constants.ZERO
+    assert await consumer.get_holdings_ratio("XYZ") == constants.ZERO
```
Comment (GuillaumeDSM) [inline tests/modes/test_abstract_mode_consumer.py:167]: 👍

File: tests/personal_data/portfolios/assets/test_future_asset.py [hunk 7]
```diff
@@ -27,45 +27,67 @@ def test___eq__():
...
+    assert asset.update(available=constants.ZERO, total=decimal.Decimal(6), unrealized_pnl=decimal.Decimal(10),
+                        position_margin=constants.ZERO, initial_margin=constants.ZERO)
+    assert asset.total == decimal.Decimal(21)
+    assert asset.wallet_balance == decimal.Decimal(11)
+    assert asset.update(available=constants.ZERO, total=decimal.Decimal(1), unrealized_pnl=decimal.Decimal(10),
+                        initial_margin=decimal.Decimal(3), position_margin=constants.ZERO)
+    assert asset.wallet_balance == decimal.Decimal(12)
+    assert asset.total == decimal.Decimal(32)
+    assert asset.initial_margin == decimal.Decimal(3)
+    assert asset.available == decimal.Decimal(12)
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/assets/test_future_asset.py:61]: 👍

File: tests/personal_data/portfolios/assets/test_future_asset.py [hunk 8]
```diff
@@ -27,45 +27,67 @@ def test___eq__():
...
+                         margin_balance=decimal.Decimal(2),
+                         initial_margin=constants.ZERO,
+                         order_margin=constants.ZERO,
+                         position_margin=constants.ZERO,
+                         unrealized_pnl=constants.ZERO)
     assert asset.available == decimal.Decimal(5)
-    assert asset.total == decimal.Decimal(5)
-    assert asset.margin_balance == decimal.Decimal(2)
+    assert asset.total == decimal.Decimal(2)
+    assert asset.wallet_balance == decimal.Decimal(5)
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/assets/test_future_asset.py:90]: should we also test invalid values to ensure edge cases are handled ?

File: tests/personal_data/portfolios/test_asset.py [hunk 9]
```diff
@@ -52,3 +52,11 @@ def test__ensure_update_validity():
         asset._ensure_update_validity(constants.ONE, -constants.ONE)
         with pytest.raises(errors.PortfolioNegativeValueError):
             asset._ensure_update_validity(-constants.ONE, constants.ZERO)
+
+def test__ensure_not_negative():
+    asset = assets.Asset(ASSET_CURRENCY_NAME, available=constants.ONE_HUNDRED, total=decimal.Decimal(2))
+    if not os.getenv('CYTHON_IGNORE'):
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/test_asset.py:58]: `_ensure_not_negative` is cpdef, it should work in cython ?

File: tests/personal_data/portfolios/types/test_future_portfolio.py [hunk 10]
```diff
@@ -14,150 +14,690 @@
...
-async def test_update_portfolio_available_from_order(backtesting_trader):
-    config, exchange_manager, trader = backtesting_trader
-    portfolio_manager = exchange_manager.exchange_personal_data.portfolio_manager
-    await init_symbol_contract(exchange_manager.exchange, leverage=decimal.Decimal(10))
+    # test buy order creation
+    portfolio_manager.portfolio.update_portfolio_available(market_buy, True)
+    assert portfolio_manager.portfolio.get_currency_portfolio("BTC").total == decimal.Decimal(10)
+    # 9 - (74 / 3) / 74 = 8.666666666666666648648648649
+    assert portfolio_manager.portfolio.get_currency_portfolio("BTC").available == decimal.Decimal(
+        '8.666666666666666648648648649')
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/types/test_future_portfolio.py:174]: this number won't make issues on different os architectures ?

File: tests/personal_data/portfolios/types/test_future_portfolio.py [hunk 11]
```diff
@@ -14,150 +14,690 @@
...
+    # Test buy order
+    market_buy = BuyMarketOrder(trader_inst)
+    market_buy.update(order_type=enums.TraderOrderType.BUY_MARKET,
+                      symbol=DEFAULT_FUTURE_SYMBOL,
+                      current_price=decimal.Decimal(str(1000)),
+                      quantity=decimal.Decimal(str(100000000)),
+                      price=decimal.Decimal(str(1000)))
+
+    # test buy order creation
+    if not os.getenv('CYTHON_IGNORE'):
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/types/test_future_portfolio.py:254]: not raised in cython ? 😮

File: tests/personal_data/portfolios/types/test_future_portfolio.py [hunk 12]
```diff
@@ -14,150 +14,690 @@
...
+    assert portfolio_manager.portfolio.get_currency_portfolio("BTC").total == decimal.Decimal(
+        str('10.10000000000000000555111512'))
+
+    position_inst.unrealised_pnl = decimal.Decimal(-0.3)
+    portfolio_manager.portfolio.update_portfolio_from_pnl(position_inst)
+    assert portfolio_manager.portfolio.get_currency_portfolio("USDT").available == decimal.Decimal(str(1000))
+    assert portfolio_manager.portfolio.get_currency_portfolio("USDT").total == decimal.Decimal(str(1000))
+    assert portfolio_manager.portfolio.get_currency_portfolio("BTC").available == decimal.Decimal(str(10))
+    assert portfolio_manager.portfolio.get_currency_portfolio("BTC").total == decimal.Decimal(
+        str('9.700000000000000011102230246'))
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/types/test_future_portfolio.py:703]: very nice tests !

File: octobot_trading/personal_data/positions/position.py [hunk 13]
```diff
@@ -414,46 +501,55 @@ def to_dict(self):
...
             enums.ExchangeConstantsPositionColumns.UNREALISED_PNL.value: self.unrealised_pnl,
             enums.ExchangeConstantsPositionColumns.REALISED_PNL.value: self.realised_pnl,
-            enums.ExchangeConstantsPositionColumns.LEVERAGE.value: self.leverage,
-            enums.ExchangeConstantsPositionColumns.MARGIN_TYPE.value: self.margin_type,
         }
 
-    async def _check_for_liquidation(self):
+    def _check_for_liquidation(self):
         """
         _check_for_liquidation will defines rules for a simulated position to be liquidated
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/personal_data/positions/position.py:508]: Yes

File: tests/personal_data/portfolios/test_asset.py (hunk 9 above)
Comment (Herklos) [reply tests/personal_data/portfolios/test_asset.py:58]: right :+1:

File: tests/personal_data/portfolios/types/test_future_portfolio.py (hunk 10 above)
Comment (Herklos) [reply tests/personal_data/portfolios/types/test_future_portfolio.py:174]: I don't know, you think it can be an issue even when using `decimal.Decimal` ?

Comment (GuillaumeDSM) [reply tests/personal_data/portfolios/types/test_future_portfolio.py:174]: Right, as long as it is initialized as a str, there should be no problem

Comment (Herklos) [reply tests/personal_data/portfolios/types/test_future_portfolio.py:174]: Right I see, I'll fix it

File: octobot_trading/personal_data/portfolios/assets/margin_asset.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        self.interest = interest
+        self.locked = locked
+
+    def __str__(self):
+        return super().__str__() + " | " \
+                                   f"Borrowed: {float(self.borrowed)} | " \
+                                   f"Interest: {float(self.interest)} | " \
+                                   f"Locked: {float(self.locked)}"
+
+    def __eq__(self, other):
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/portfolios/assets/margin_asset.py:36]: 👍

File: octobot_trading/personal_data/positions/channel/positions_updater_simulator.py [hunk 15]
```diff
@@ -38,45 +36,39 @@ async def initialize(self) -> None:
...
         try:
-            position.mark_price = mark_price
-            await position.update_position_status()
-
-            if position.is_liquidated():
-                self.logger.debug(f"{position.symbol} (ID : {position.position_id})"
-                                  f" liquidated on {self.channel.exchange.name} at {position.mark_price} ")
+            for symbol_position in self.channel.exchange_manager.exchange_personal_data.positions_manager. \
+                    get_symbol_positions(symbol=symbol):
+                if not symbol_position.is_idle() and symbol_position.symbol_contract.is_perpetual_contract():
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/positions/channel/positions_updater_simulator.py:68]: 👍

File: octobot_trading/personal_data/positions/positions_manager.py [hunk 16]
```diff
@@ -89,6 +113,20 @@ def clear(self):
...
+        :param symbol: the position symbol
+        :param side: the position side
+        :param contract: the symbol contract (optional)
+        :return: the computed position id
+        """
+        symbol_contract = self.trader.exchange_manager.exchange.get_pair_future_contract(symbol) \
+            if contract is None else contract
+        if symbol_contract.is_one_way_position_mode():
+            return symbol
+        return f"{symbol}{self.POSITION_ID_SEPARATOR}{side.value}"
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/positions/positions_manager.py:128]: 👍

File: octobot_trading/personal_data/positions/types/inverse_position.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        try:
+            self.value = self.quantity / self.mark_price
+        except (decimal.DivisionByZero, decimal.InvalidOperation):
+            self.value = constants.ZERO
+
+    def update_pnl(self):
+        """
+        LONG_PNL = CONTRACT_QUANTITY x [(1 / ENTRY_PRICE) - (1 / MARK_PRICE)]
+        SHORT_PNL = CONTRACT_QUANTITY x [(1 / MARK_PRICE) - (1 / ENTRY_PRICE)]
```
Comment (GuillaumeDSM) [inline octobot_trading/personal_data/positions/types/inverse_position.py:35]: 👍

File: tests/personal_data/portfolios/test_asset.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    asset = assets.Asset(ASSET_CURRENCY_NAME, available=constants.ONE_HUNDRED, total=constants.ONE)
+    assert asset.to_dict() == {
+        common_constants.PORTFOLIO_AVAILABLE: constants.ONE_HUNDRED,
+        common_constants.PORTFOLIO_TOTAL: constants.ONE
+    }
+
+
+def test__ensure_update_validity():
+    asset = assets.Asset(ASSET_CURRENCY_NAME, available=constants.ONE_HUNDRED, total=decimal.Decimal(2))
+    if not os.getenv('CYTHON_IGNORE'):
```
Comment (GuillaumeDSM) [inline tests/personal_data/portfolios/test_asset.py:46]: we really need to fix this ^^

File: tests/personal_data/positions/test_positions_manager.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+
+async def test__generate_position_id(future_trader_simulator_with_default_linear):
+    config, exchange_manager, trader, default_contract = future_trader_simulator_with_default_linear
+    positions_manager = exchange_manager.exchange_personal_data.positions_manager
+    symbol_contract = default_contract
+    trader.exchange_manager.exchange.set_pair_future_contract(DEFAULT_FUTURE_SYMBOL, 
+                                                              default_contract)
+
+    if not os.getenv('CYTHON_IGNORE'):
```
Comment (GuillaumeDSM) [inline tests/personal_data/positions/test_positions_manager.py:99]: what is the cython issue here ?

**Review by GuillaumeDSM — COMMENTED** [review]
🎉 🎉 🎉

File: octobot_trading/exchanges/implementations/future_ccxt_exchange.py [hunk 20]
```diff
@@ -167,3 +168,64 @@ def clean_trade(self, trade):
...
+
+    async def set_symbol_leverage(self, symbol: str, leverage: int):
+        return await self.connector.set_symbol_leverage(leverage=leverage, symbol=symbol)
+
+    async def set_symbol_margin_type(self, symbol: str, isolated: bool):
+        return await self.connector.set_symbol_margin_type(symbol=symbol, isolated=isolated)
+
+    def parse_position(self, position_dict) -> dict:
+        try:
+            position_dict.update({
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/exchanges/implementations/future_ccxt_exchange.py:189]: Yes, that's right, I'll create a new enum dedicated to ccxt keys

File: octobot_trading/personal_data/orders/order.py [hunk 21]
```diff
@@ -197,6 +203,12 @@ def update(self, symbol, order_id="", status=enums.OrderStatus.OPEN,
         if not self.reduce_only:
             self.reduce_only = reduce_only
 
+        if not self.close_position:
+            self.close_position = close_position
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/personal_data/orders/order.py:207]: Hum, strange condition, I'll have a look into it

File: tests/personal_data/positions/test_positions_manager.py (hunk 19 above)
Comment (Herklos) [reply tests/personal_data/positions/test_positions_manager.py:99]: _generate_position_id is a `cdef`

File: octobot_trading/exchanges/implementations/future_ccxt_exchange.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/exchanges/implementations/future_ccxt_exchange.py:189]: Good point but it's already done, there is `ExchangeConstantsPositionColumns` and `ExchangePositionCCXTColumns` or I miss something

**Review by GuillaumeDSM — APPROVED** [review]
🎉 🌮 🚀 🐙
